### PR TITLE
Add support for key hashing

### PR DESCRIPTION
**Pull request by akirk**
This adds support for compressed jed as defined in D10012-code. The `key-hash` in the header specifies a hash algorithm (currently only sha1 supported) and it's length, so `sha1-4` would mean that only the first 4 characters of the hex representation of the original string should be used for lookup.

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -20,6 +21,12 @@ var numberFormatPHPJS = require( './number-format' );
 var decimal_point_translation_key = 'number_format_decimals',
 	thousands_sep_translation_key = 'number_format_thousands_sep';
 
+var hashers = [
+	function( options ) {
+		return options;
+	}
```
Comment (ramonjd) [inline lib/index.js]: What does the function at `0` do? Does it deserve a comment or should it be obvious?

File: lib/index.js [hunk 2]
```diff
@@ -167,6 +174,38 @@ I18N.prototype.configure = function( options ) {
...
 
 I18N.prototype.setLocale = function( localeData ) {
+	if ( localeData && localeData[ '' ] && localeData[ '' ][ 'key-hash' ] ) {
+		var hash, hashLength, keyHash = localeData[ '' ][ 'key-hash' ];
+
+		if ( keyHash.substr( 0, 4 ) === 'sha1' ) {
+			hashLength = Number ( keyHash.substr( 5 ) );
+			hash = function( string ) {
+				var hash = sha1();
+				hash.update( string );
```
Comment (ramonjd) [inline lib/index.js]: Using a `var` name that appears in a higher scope could be confusing at first to those who don't know about javascript function contexts.

File: lib/index.js [hunk 3]
```diff
@@ -167,6 +174,38 @@ I18N.prototype.configure = function( options ) {
...
+				hash = hash.digest( 'hex' );
```
Comment (ramonjd) [inline lib/index.js]: I can't work out why, but `hash.js` returns a different hash for ellipses. 

Here's a comparison with the [sha1](https://www.npmjs.com/package/sha1) lib.

```
sha1( '…' ) sha1 df316e17de9c36858417ba18e0da4da282dea84a
index.jsx:81 hash.sha1().update('…').digest('hex') hash.js sha1 552cb47dfe2134b226b093b3fa5df3f381984235
```

I've noticed it on the following string:

```
sha1( 'Search all receipts…' ) // 5c4e5bd997b566add1bbc776b546fe6d18f4abe0
hash.sha1().update('Search all receipts…').digest('hex') // 7614de0bfd4652d8f72906c55f0bed4bc5f87b0c
```

screen shot 2018-02-26 at 3 08 04 pm

I haven't tested it with other character entities.

**Comment by ramonjd** [comment]
For some items in the nav, the translations don't after a page refresh, only switching to German when the items rerender (triggered by clicking around).

feb-26-2018 16-38-53

File: lib/index.js (hunk 1 above)
Comment (akirk) [reply lib/index.js]: Thanks, I added a comment.

File: lib/index.js (hunk 2 above)
Comment (akirk) [reply lib/index.js]: Good call, changed.

File: lib/index.js (hunk 3 above)
Comment (akirk) [reply lib/index.js]: That's worrying, thanks for noticing! I have changed the dependency to `sha1` and will do the same for Calypso (which was reason in the first place I opted for hash.js).

**Comment by akirk** [comment]
I see this behavior already in the current master, that piece of the sidebar is implemented in a slightly strange way.

**Comment by ramonjd** [comment]
> I see this behavior already in the current master, that piece of the sidebar is implemented in a slightly strange way.

Cool. I should have mentioned that I was running sb. Might be laaaaaag. 👍

File: lib/index.js [hunk 4]
```diff
@@ -243,6 +316,17 @@ I18N.prototype.addTranslations = function( localeData ) {
 };
 
 
+/**
+ * Checks whether the given original has a translation.
+ *
+ * @param  {string} original  - the string to translate
```
Comment (akirk) [inline lib/index.js]: TODO: fix doc header.

File: lib/index.js [hunk 5]
```diff
@@ -252,22 +336,25 @@ I18N.prototype.addTranslations = function( localeData ) {
  * @return {string|React-components} translated text or an object containing React children that can be inserted into a parent component
  */
 I18N.prototype.translate = function() {
-	var options, translation, sprintfArgs, errorMethod, optionsString, cacheable;
+	var options, optionsString, translation, sprintfArgs, errorMethod, cacheable, i;
```
Comment (akirk) [inline lib/index.js]: Note to self: revert this.

**Pull request merged**